File: analytical.R
#importing libraries
library("tidyr")
library("devtools")
library("dplyr")

#loading data
loans = read.csv('loans_r.csv') #loading data in the variable loans
accounts = read.csv('data/accounts.csv')
links = read.csv('data/links.csv')
transactions = read.csv('data/transactions.csv')
payment_orders = read.csv('data/payment_orders.csv')
cards = read.csv('data/cards.csv')
districts = read.csv('district_r.csv')


df = accounts #df will be the main dataframe that I will output

colnames(df) <- c('account_id', 'id','open_date','statement_frequency') 

df <- merge(x=df,y=districts[,c("id","name")],by="id",all.x=TRUE) #left join

df <- subset(df, select = -c(id)) #drop column id

df <- df[,c(1,4,2,3)] # reorder columns
colnames(df) <- c('account_id', 'district_name','open_date','statement_frequency') #rename columns

df <- merge(x=df,y=links[,c("account_id","client_id")],by="account_id",all.x=TRUE) #left join

c_customers <- df %>% count(account_id)  # Applying count to count the num of clients with each account
colnames(c_customers) <- c('account_id', 'num_customers') #rename columns

df <- df[!duplicated(df$account_id),] #removing duplicated account_ids
df <- merge(x=df,y=c_customers[,c("account_id","num_customers")],by="account_id",all.x=TRUE) #left join
df <- subset(df, select = -c(client_id)) #drop client id columns

#cards.csv
colnames(cards) <- c('id','client_id','type','issue_date') #changing the names of the columns

ccard = merge(x=links,y=cards[,c("client_id","id")],by="client_id") #inner join 
c_ccard <- ccard %>% count(account_id) #counts the number of credit cards per account_id
colnames(c_ccard) <- c('account_id','credit_cards')
df <- merge(x=df,y=c_ccard[,c("account_id","credit_cards")],by="account_id",all.x=TRUE)
df <- df %>% mutate(credit_cards = ifelse(is.na(credit_cards), 0, credit_cards)) #replace na with 0

#loans.csv

colnames(loans) <- c('id','account_id','date', 'loan_amount', 'loan_payments','loan_term','loan_status', 'loan_default')
loans$loan = TRUE #make a new column called loans with all values True
df <- merge(x=df,y=loans[,c("account_id",'loan_amount', 'loan_payments',
                            'loan_term','loan_status', 'loan_default','loan')]
                         ,by ="account_id",all.x=TRUE)

df <- df %>% mutate(loan = ifelse(is.na(loan), FALSE, loan)) #replace na with FALSE

#transactions.csv

withdrawal <- transactions[which(transactions$type=='debit'),] #data for all withdrawals
withdrawal <- subset(withdrawal, select = c(account_id,amount)) #only two columns are needed hence subsetting
c_withdrawal_max <- aggregate(amount~account_id, withdrawal,max)#finding max withdrawal of each account_id
c_withdrawal_min <- aggregate(amount~account_id, withdrawal,min)#finding min withdrawal of each account_id
colnames(c_withdrawal_max) <-c("account_id", "max_withdrawal")#changing names of columns
colnames(c_withdrawal_min) <-c("account_id", "min_withdrawal")#changing names of columns
c_withdrawal <- merge(x=c_withdrawal_max,y=c_withdrawal_min
            ,by ="account_id")# inner join 
df<-merge(x=df,y=c_withdrawal,by ="account_id",all.x=TRUE) #joining to main df


ccpayments <- transactions[which(transactions$method=='credit card'),]
ccpayments <- ccpayments[which(ccpayments$type=='debit'),]
ccpayments <- subset(ccpayments, select = c(account_id))
c_ccpayments <- ccpayments %>% count(account_id)
colnames(c_ccpayments) <-c("account_id","cc_payments")
df<-merge(x=df,y=c_ccpayments,by ="account_id",all.x=TRUE) #joining to main df
df <- df %>% mutate(cc_payments = ifelse(is.na(cc_payments), 0, cc_payments))#changing na values to 0

bal <- subset(transactions, select = c(account_id,balance))#subsetting since we only need the two columns
c_bal_max <- aggregate(balance~account_id, bal,max)#storing max balance in the account here
c_bal_min <- aggregate(balance~account_id, bal,min)#storing min balance in the account here
colnames(c_bal_max) <-c("account_id", "max_balance")#changing names of columns
colnames(c_bal_min) <-c("account_id", "min_balance")#changing names of columns
c_bal <- merge(x=c_bal_max,y=c_bal_min,by ="account_id")# inner join 
df<-merge(x=df,y=c_bal,by ="account_id",all.x=TRUE) #left outer joining to main df

df <- df[,c(1,2,3,4,5,6,12,7,8,9,10,11,13,14,15,16,17)] #reorder columns

#converting the modified data to a csv 
write.csv(df, "analytical_r.csv", row.names=FALSE) #done yay

File: analytical.py
import pandas as pd
loans = pd.read_csv('loans_py.csv') #loading data in the variable loans
accounts = pd.read_csv('data/accounts.csv')
links = pd.read_csv('data/links.csv')
transactions = pd.read_csv('data/transactions.csv')
payment_orders = pd.read_csv('data/payment_orders.csv')
cards = pd.read_csv('data/cards.csv')
districts = pd.read_csv('district_py.csv')


df = accounts #df is the main dataframe that will be outputted, using data from accounts.csv to put in df

df.columns = ['account_id', 'id','open_date','statement_frequency']

df = pd.merge(df, districts[["id", "name"]], how="left", on=["id"]) #left outer join on df and districts
df = df.drop(['id'], axis=1) #drop column id
df = df[['account_id','name','open_date','statement_frequency']] #reorder columns
df.columns = ['account_id', 'district_name','open_date','statement_frequency'] #change names of columns


df = pd.merge(df, links[["account_id", "client_id"]], how="left", on=["account_id"]) #left join
s = df.groupby('account_id').size().rename_axis('account_id').reset_index(name='num_customers') #makes a df called s that has number of customers linked to an account

df = df.drop_duplicates('account_id')

#s.columns = ['account_id', 'num_customers']
df = pd.merge(df, s[["account_id","num_customers"]], how="left", on=["account_id"]) #joining this df created to the main df



cards.columns = ['id','client_id','type','issue_date'] #changing the names of the columns
ccard = pd.merge(links, cards[["client_id",'id']], how="inner", on=["client_id"])  #inner join to link the card info to links.csv
ccard = ccard.groupby('account_id').size().rename_axis('account_id').reset_index(name='credit_cards')# makes a df called ccard that has number of cards linked to an account
df = pd.merge(df, ccard[["account_id",'credit_cards']], how="left", on=["account_id"]) #left join since not all customers have cards
df['credit_cards'] = df['credit_cards'].fillna(0) #replacing the na values with 0


loans.columns = ['id','account_id','date', 'loan_amount', 'loan_payments','loan_term','loan_status', 'loan_default'] #changing names of loans_py.csv for convenience
loans['loan'] = True #making a new column that has true for all customers that have or have had a loan
df = pd.merge(df, loans[["account_id",'loan_amount', 'loan_payments','loan_term','loan_status', 'loan_default','loan']], how="left", on=["account_id"]) # left join since not all accounts have a loan
df["loan"].fillna(False,inplace=True) #false inplace of na

withdrawal = transactions[transactions['type']=='debit'] #data for all withdrawals
withdrawal = withdrawal[['account_id','amount']] #changing col names
max_withdrawal = withdrawal.loc[withdrawal.groupby('account_id')['amount'].idxmax(), :].reset_index() #creating a df which has the max withdrawal for each account_id
max_withdrawal.columns = ['index','account_id','max_withdrawal']#changing names of the created df
min_withdrawal = withdrawal.loc[withdrawal.groupby('account_id')['amount'].idxmin(), :].reset_index()#creating a df which has the min withdrawal for each account_id
min_withdrawal.columns = ['index','account_id','min_withdrawal']#changing names of the created df
df = pd.merge(df,max_withdrawal[['account_id','max_withdrawal']],how="left", on=["account_id"])
df = pd.merge(df,min_withdrawal[['account_id','min_withdrawal']],how="left", on=["account_id"])#left outer join this info

ccpayments = transactions[(transactions['method']=='credit card')] #data for all credit card transactions
ccpayments = ccpayments[ccpayments['type']=='debit']
ccpayments = ccpayments[['account_id']] #we only need account_id to count
ccpayments = ccpayments.groupby('account_id').size().rename_axis('account_id').reset_index(name='cc_payments') #making a new column that stores the number of transactions
df = pd.merge(df, ccpayments[["account_id",'cc_payments']], how="left", on=["account_id"])#left join
df['cc_payments'] = df['cc_payments'].fillna(0) #i dont know if this should be left as NA or 0


bal = transactions[['account_id','balance']] #looking at the balances and account_id only
max_bal = bal.loc[bal.groupby('account_id')['balance'].idxmax(), :].reset_index()  #stores the max balance for every account_id
min_bal = bal.loc[bal.groupby('account_id')['balance'].idxmin(), :].reset_index() #stores the min balance for every account_id
max_bal = max_bal.drop(['index'], axis = 1)#drop index column
min_bal = min_bal.drop(['index'], axis = 1)

tot_bal = pd.concat([max_bal,min_bal], axis = 1, join = "inner") #creates a new df with both max and min balances and account_d
tot_bal.columns = ['account_id','max_balance','drop','min_balance'] #renames
tot_bal = tot_bal.drop(['drop'], axis=1) #drops the extra account_id

df = pd.merge(df, tot_bal[["account_id",'max_balance','min_balance']], how="left", on=["account_id"])#adding this to the main df

del df['client_id'] #extra info not required
df = df[['account_id', 'district_name', 'open_date', 'statement_frequency',
       'num_customers', 'credit_cards', 'loan', 'loan_amount',
       'loan_payments', 'loan_term', 'loan_status', 'loan_default',
       'max_withdrawal', 'min_withdrawal', 'cc_payments', 'max_balance',
       'min_balance']] #changed the order to the required


df.to_csv('analytical_py.csv', index=False, encoding='utf-8') #convert to csv and done yay